plugins/ipam/host-local/backend/allocator/allocator.go

```diff
@@ -151,39 +150,67 @@ func (a *IPAllocator) GetIter() (*RangeIter, error) {
 	if err != nil && !os.IsNotExist(err) {
 		log.Printf("Error retrieving last reserved ip: %v", err)
 	} else if lastReservedIP != nil {
-		startFromLastReservedIP = a.ipRange.IPInRange(lastReservedIP) == nil
+		startFromLastReservedIP = a.pool.Contains(lastReservedIP) == nil
 	}
 
+	// Find the range in the pool with this IP
 	if startFromLastReservedIP {
-		if i.high.Equal(lastReservedIP) {
-			i.start = i.low
-		} else {
-			i.start = ip.NextIP(lastReservedIP)
+		for i, r := range *a.pool {
+			if r.Contains(lastReservedIP) == nil {
+				iter.rangeIdx = i
+				iter.startRange = i
+
+				// We advance the cursor on ever Next(), so the first call
+				// to next() will return lastReservedIP + 1
+				iter.cur = lastReservedIP
+				break
+			}
 		}
 	} else {
-		i.start = a.ipRange.RangeStart
+		iter.rangeIdx = 0
+		iter.startRange = 0
+		iter.startIP = (*a.pool)[0].RangeStart
 	}
-	return &i, nil
+	return &iter, nil
 }
 
 // Next returns the next IP in the iterator, or nil if end is reached
-func (i *RangeIter) Next() net.IP {
-	// If we're at the beginning, time to start
+func (i *RangeIter) Next() (*net.IPNet, net.IP) {
+	r := (*i.pool)[i.rangeIdx]
+
+	// If this is the first time iterating and we're not starting in the middle
+	// of the range, then start at rangeStart, which is inclusive
 	if i.cur == nil {
-		i.cur = i.start
-		return i.cur
+		i.cur = r.RangeStart
+		i.startIP = i.cur
+		if i.cur.Equal(r.Gateway) {
+			return i.Next()
+		}
+		return &net.IPNet{i.cur, r.Subnet.Mask}, r.Gateway
 	}
-	//  we returned .high last time, since we're inclusive
-	if i.cur.Equal(i.high) {
-		i.cur = i.low
+
+	// If we've reached the end of this range, we need to advance the range
+	// RangeEnd is inclusive as well
+	if i.cur.Equal(r.RangeEnd) {
+		i.rangeIdx += 1
+		i.rangeIdx %= len(*i.pool)
+		r = (*i.pool)[i.rangeIdx]
```

Comment:
Pool exhaustion and end-of-range are not the same, since we can start in the "middle" of the range set. This code wont out-of-range since we wrap it back around with a modulo.

During Canonicalize, we reject empty pools.